geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/ThreadManagement.scala

```diff
@@ -28,28 +36,120 @@ object ThreadManagement extends LazyLogging {
   }
 
   /**
-   * Register a query with the thread manager
+   * Register a scan with the thread manager
+   *
+   * @param scan scan to terminate
+   * @return
    */
-  def register(query: ManagedQuery): ScheduledFuture[_] =
-    executor.schedule(new QueryKiller(query), query.getTimeout, TimeUnit.MILLISECONDS)
+  def register(scan: ManagedScan[_]): ScheduledFuture[_] = {
+    val timeout = math.max(1, scan.timeout.absolute - System.currentTimeMillis())
+    executor.schedule(new QueryKiller(scan), timeout, TimeUnit.MILLISECONDS)
+  }
 
   /**
-    * Trait for classes to be managed for timeouts
-    */
-  trait ManagedQuery extends Closeable {
-    def getTimeout: Long
-    def isClosed: Boolean
-    def debug: String
-  }
+   * Trait for scans that are managed, i.e. tracked and terminated if they exceed a specified timeout
+   *
+   * @tparam T type
+   */
+  trait ManagedScan[T] extends CloseableIterator[T] {
+
+    /**
+     * Scan timeout
+     *
+     * @return
+     */
+    def timeout: Timeout
+
+    /**
+     * Low-level scan to be stopped
+     *
+     * @return
+     */
+    protected def underlying: LowLevelScanner[T]
+
+    // used for log messages
+    protected def typeName: String
+    protected def filter: Option[Filter]
+
+    // we can use a volatile var since we only update the value with a single thread
+    @volatile
+    private var terminated = timeout.absolute <= System.currentTimeMillis()
+
+    private val iter = ExceptionalIterator(if (terminated) { Iterator.empty } else { underlying.iterator })
+    private val cancel = if (terminated) { None } else { Some(ThreadManagement.register(this)) }
```

Comment:
Ok, just to make sure I'm understanding... this is called as part of the constructor / initialization for the ManagedScan.  So if the Scan is initialized *late*, then it is already terminated, etc, etc, etc...

Reply:
correct